### PR TITLE
feat: tx valid since

### PR DESCRIPTION
**Pull request by jjyr**

### CODE REVIEW

File: shared/src/shared.rs [hunk 1]
```diff
@@ -81,13 +82,14 @@ impl<CI: ChainIndex> Shared<CI> {
                 proposal_ids,
                 TxPool::new(tx_pool_config),
                 LruCache::new(txs_verify_cache_size),
+                Arc::clone(&consensus),
```
Comment (quake) [inline shared/src/shared.rs]: I think that we can refactor ChainState construct function now:
```rust
ChainState::new(&store, tx_pool_config, Arc::clone(&consensus))
```

and move all init code to construct function.

File: verification/src/transaction_verifier.rs [hunk 2]
```diff
@@ -25,6 +37,7 @@ impl<'a> TransactionVerifier<'a> {
             script: ScriptVerifier::new(rtx),
             capacity: CapacityVerifier::new(rtx),
             inputs: InputVerifier::new(rtx),
+            valid_since: ValidSinceVerifier::new(&rtx.transaction, &block_context),
```
Comment (u2) [inline verification/src/transaction_verifier.rs]: Call `self.valid_since.verify()?;` in `verify`.

https://github.com/nervosnetwork/ckb/pull/372/files#diff-d0ea0a76c2e66e05fbcf6f957b6eccf4R44

**Comment by jjyr** [comment]
bors try

**Comment by jjyr** [comment]
RFC update https://github.com/nervosnetwork/rfcs/pull/97

File: shared/src/chain_state.rs [hunk 3]
```diff
@@ -39,14 +39,35 @@ impl<CI: ChainIndex> ChainState<CI> {
...
     pub fn new(
         store: &Arc<CI>,
-        tip_header: Header,
-        total_difficulty: U256,
-        cell_set: CellSet,
-        proposal_ids: TxProposalTable,
-        tx_pool: TxPool,
-        txs_verify_cache: LruCache<H256, Cycle>,
         consensus: Arc<Consensus>,
+        txs_verify_cache_size: usize,
```
Comment (quake) [inline shared/src/chain_state.rs]: `txs_verify_cache_size` is one field of `TxPoolConfig`, we can omit it here.

File: shared/src/chain_state.rs [hunk 4]
```diff
@@ -39,14 +39,35 @@ impl<CI: ChainIndex> ChainState<CI> {
     #[allow(clippy::too_many_arguments)]
```
Comment (quake) [inline shared/src/chain_state.rs]: Please remove this clippy hint

File: shared/src/chain_state.rs (hunk 3 above)
Comment (jjyr) [reply shared/src/chain_state.rs]: Logically `txs_verify_cache_size` is not related with `TxPoolConfig`, you can see we use `txs_verify_cache` in `chain` or other modules, it is more like a shared object.

I'd rather leave it here for now.

Comment (quake) [reply shared/src/chain_state.rs]: Please check this line:
https://github.com/nervosnetwork/ckb/blob/61855c678d7806e5dc733e3806757d2f47a0a47b/shared/src/shared.rs#L305

`txs_verify_cache_size` is copied from `TxPoolConfig`

Comment (jjyr) [reply shared/src/chain_state.rs]: Fair enough

File: chain/src/chain.rs [hunk 5]
```diff
@@ -191,10 +225,10 @@ impl<CI: ChainIndex + 'static> ChainService<CI> {
         let current_total_difficulty = chain_state.total_difficulty().clone();
 
         debug!(
-            target: "chain",
-            "difficulty current = {}, cannon = {}",
-            current_total_difficulty,
-            cannon_total_difficulty,
+        target: "chain",
```
Comment (zhangsoledad) [inline chain/src/chain.rs]: meaningless indent

**Comment by zhangsoledad** [comment]
The transaction with a future `valid_since`: 
* RPC:  be discarded directly
* Relay:  staging failed and then be discarded.

Any plan for this?

**Comment by jjyr** [comment]
> The transaction with a future `valid_since`:
> 
> * RPC:  be discarded directly
> * Relay:  staging failed and then be discarded.
> 
> Any plan for this?

We just reject to send immature tx in RPC, which is bitcoin currently behavior.

File: core/src/transaction.rs [hunk 6]
```diff
@@ -63,32 +63,36 @@ impl OutPoint {
...
+    pub fn new(previous_output: OutPoint, valid_since: u64, args: Vec<Vec<u8>>) -> Self {
         CellInput {
             previous_output,
+            valid_since,
             args,
         }
     }
 
-    pub fn new_cellbase_input(block_number: BlockNumber) -> Self {
+    pub fn new_cellbase_input(block_number: BlockNumber, valid_since: u64) -> Self {
```
Comment (quake) [inline core/src/transaction.rs]: I think that `valid_since` of Cellbase should always be 0, no need to pass the parameter here.

Comment (jjyr) [reply core/src/transaction.rs]: Fixed

File: verification/src/transaction_verifier.rs [hunk 7]
```diff
@@ -209,3 +225,161 @@ impl<'a> CapacityVerifier<'a> {
...
     }
 }
+
+const LOCK_TYPE_FLAG: u64 = 1 << 63;
+const TIME_TYPE_FLAG: u64 = 1 << 62;
+const VALUE_MUSK: u64 = 0x00ff_ffff_ffff_ffff;
+
+/// RFC 0017
+#[derive(Copy, Clone, Debug)]
+struct ValidSince(u64);
```
Comment (quake) [inline verification/src/transaction_verifier.rs]: I prefer enum to struct:
```rust
enum ValidSince {
  None,
  AbsoluteNumber(u64),
  AbsoluteTimestamp(u64),
  RelativeNumber(u64),
  RelativeTimestamp(u64),
}
```
And impl From/To u64 trait, that will remove these helper methods and simplify code.

Comment (jjyr) [reply verification/src/transaction_verifier.rs]: tomayto tomato, I can't see big improvement, it just change the conditions check from function to pattern match,

Comment (quake) [reply verification/src/transaction_verifier.rs]: It's a common refactoring practice in other OOP language, it is called `Replace Conditional with Polymorphism`.

Comment (jjyr) [reply verification/src/transaction_verifier.rs]: Can't agree, If we support more bits flags, the enum will increment exponentially, there's no clear improvement.

Comment (jjyr) [reply verification/src/transaction_verifier.rs]: Do not have a conclusion, just use the current implementation for now.

**Comment by doitian** [comment]
https://travis-ci.com/nervosnetwork/ckb/builds/107753855

File: protocol/src/protocol_generated_verifier.rs [hunk 8]
```diff
@@ -52,19 +52,14 @@ pub mod ckb {
 
                 let vtab = tab.vtable();
                 let vtab_num_bytes = vtab.num_bytes();
-                let object_inline_num_bytes = vtab.object_inline_num_bytes();
```
Comment (doitian) [inline protocol/src/protocol_generated_verifier.rs]: This is generated by an old version of cfbc, please install 0.19 and regenerate it

**Pull request merged**